Adds GA tracking

elenatanasoiu opened the pull request:
[Trello card](https://trello.com/c/GGvHKuad/118-event-tracking-and-analytics-for-iteration-2-of-local-links-manager-3)

Depends on changes being merged in https://github.com/alphagov/govuk_admin_template/pull/145 This will enable modules defined in govuk_admin_template to be used in local-links-manager.

We send out events (`track-submit`, `track-click`, `auto-track-event` etc) to GA.

fofr commented on app/views/local_authorities/index.html.erb:
```diff
@@ -15,7 +15,10 @@
-        <div class='count'>
+        <div class='count'
+             data-module="track-dimension"
+             data-track-index="Scorecard total",
```
Index should be a number.

elenatanasoiu replied:
Changed to number

Commit:
Enables GA tracking

(Trello card)[https://trello.com/c/GGvHKuad/118-event-tracking-and-analytics-for-iteration-2-of-local-links-manager-3]

Depends on changes being merged in alphagov/govuk_admin_template#145

This will enable modules defined in govuk_admin_template to be used in local-links-manager.
With these we can start sending events to Google Analytics.

boffbowsh commented on Gemfile:
```diff
@@ -12,7 +12,7 @@ gem 'airbrake', '~> 4.0'
 gem 'faraday'
 gem 'faraday_middleware'
-gem 'govuk_admin_template', '~> 4.2'
+gem 'govuk_admin_template'
```
This should now be `~> 5.0`

elenatanasoiu replied:
Added

boffbowsh reviewed (changes requested):
Couple of small changes. This looks mostly good though - can you get a review from @fofr please?

elenatanasoiu commented:
Asked @fofr for review

fofr commented on spec/javascripts/support/jasmine_helper.rb:
```diff
@@ -0,0 +1,16 @@
+# Example: prevent PhantomJS auto install, uses PhantomJS already on your path.
+Jasmine.configure do |config|
+  if ENV['TRAVIS']
+    config.prevent_phantom_js_auto_install = true
```
Why is this limited to `if ENV['TRAVIS']` and not applied all the time?

Other apps have this config set all the time:
eg https://github.com/alphagov/govuk_admin_template/blob/master/spec/javascripts/support/jasmine_helper.rb

elenatanasoiu replied:
I've removed the environment check

Commit:
Adds search tracking for index pages

This will fire a `filterUsed` event when a user writes something in the filter search box
(action: `filterUsed`, category: `userInteraction:LLM`)

This commit also sets a custom dimension to be sent along with the GA events being sent in
LocalLinksManager (custom dimension index: 7, name: `Scorecard Total`). The Scorecard total
is a value that indicates the number of links being displayed on an index page. It is sent
when the index pages first loads, using the `auto_track_event` module.

fofr commented on app/assets/javascripts/application.js:
```diff
@@ -0,0 +1,3 @@
+//= require jquery
+//= require jquery_ujs
```
govuk_admin_template provides jQuery so the top line can be removed.
What is `jquery_ujs` used for?

elenatanasoiu replied:
I've removed them. I believe I was using them to check jasmine tests in the browser, but they are redundant, I agree.

fofr commented on app/assets/javascripts/modules/track_search.js:
```diff
@@ -0,0 +1,24 @@
+
+      var filterSearchInput = container.find('table'),
+          tableInput = container.find('.js-filter-table-input');
+      filterSearchInput.on('keyup', '.js-filter-table-input', trackSearch);
```
This will fire an analytics event after every keypress. What's the intention of this tracking? What are you trying to  find out?

elenatanasoiu replied:
Ideally it's to send what's being searched back to GA. However, we could just fire that the filter was being used. 

I'll wait to talk to Tim LB before continuing with this, as it might be dropped.

fofr commented on app/views/interactions/index.html.erb:
```diff
@@ -43,7 +43,13 @@
         </td>
         <td class='text-center'>
-          <%= link_to(interaction.button_text, edit_link_path(interaction_slug: interaction.slug), class: "btn btn-default btn-s") %>
+          <div class='actions add-vertical-margins col-xs-6 no-gutter'
```
Minor: inconsistent attribute quotes

elenatanasoiu replied:
Fixed wherever I could find it.

fofr commented on app/views/interactions/index.html.erb:
```diff
@@ -43,7 +43,13 @@
+               data-track-action="editButtonClick"
+               data-track-label="<%= edit_link_path(interaction_slug: interaction.slug) %>"
+               data-track-category="userInteraction:LLM">
+               <%= link_to(interaction.button_text, edit_link_path(interaction_slug: interaction.slug), class: "btn btn-default btn-s js-track") %>
```
I'm not sure this will track as you intended. The module pulls the label and value from the link, and the action and category from the container. This will use the text of the link as the label, rather than the one provided in the attribute.

See: https://github.com/alphagov/govuk_admin_template/pull/145/files#diff-3018827fa99e362685623cba2c52c238

What are the styles on the wrapping `div` necessary for? If the layout has changed, could you include screenshots?

elenatanasoiu replied:
I've removed the div wrapping. And I've moved the label as a property on the link.

elenatanasoiu replied:
This is what the div wrapping was for (red Delete button)
<img width="1216" alt="screen shot 2017-04-12 at 17 49 36" src="https://cloud.githubusercontent.com/assets/1354439/24969647/5402f1a6-1fa9-11e7-911f-2d224fe51fc7.png">

This wasn't necessary on the index pages which looks like this (Edit button)
<img width="1243" alt="screen shot 2017-04-12 at 17 56 48" src="https://cloud.githubusercontent.com/assets/1354439/24969671/6ef0572e-1fa9-11e7-94ab-f133e74a8ffd.png">

fofr commented on app/views/links/edit.html.erb:
```diff
@@ -1,38 +1,53 @@
 
-<div class="page-title">
-  <h1><%= @local_authority.name %></h1>
+  <div class="page-title">
```
The new indentation here is wrong.

elenatanasoiu replied:
Fixed

fofr commented on app/views/local_authorities/index.html.erb:
```diff
@@ -8,30 +8,41 @@
-        </td>
+
+          <div class="count">
+            <% content_for :custom_dimension_js do %>
```
Why is this template variable being set here? I'd expect it to be set somewhere further up, at the start of a template or elsewhere.

elenatanasoiu replied:
I've moved this up to the top of the page

fofr commented on spec/javascripts/support/jasmine_helper.rb:
```diff
@@ -0,0 +1,14 @@
+#
+# Example: prevent PhantomJS auto install, uses PhantomJS already on your path.
+Jasmine.configure do |config|
+  config.prevent_phantom_js_auto_install = true
```
:+1:

fofr commented on app/assets/javascripts/modules/track_search.js:
```diff
@@ -0,0 +1,24 @@
+
+    that.start = function(container) {
+
+      var filterSearchInput = container.find('table'),
```
A table is not a`filterSearchInput`.

The two var declarations and the event listener can be reduced to:
`container.on('keyup', '.js-filter-table-input', trackSearch);`

elenatanasoiu replied:
I've renamed tableInput to filterSearchInput and removed the original var: 
`filterSearchInput = container.find('.js-filter-table-input');`

I've replaced `filterSearchInput.on('keyup', '.js-filter-table-input', trackSearch);` with `container.on('keyup', '.js-filter-table-input', trackSearch);`

However, I still need the new `filterSearchInput` in order to get the text being searched for, so I've kept that variable.

fofr commented on app/assets/javascripts/modules/track_search.js:
```diff
@@ -0,0 +1,24 @@
+        var action = container.data('track-action'),
+          label = container.data('track-label'),
+          category = container.data('track-category'),
+          value = $.trim(tableInput.val());
```
Values must be numerical, this won't work.

elenatanasoiu replied:
I've moved the text being search to the label variable.

fofr reviewed (changes requested):
I've found some significant problems with this PR. Specifically, it's not clear what the filter input analytics intends to track – it's sending a lot of events and the data attributes passed to it contain mistakes.

Was this tested before the PR was raised?

Commit:
Adds jasmine via official jasmine gem

Commit:
Adds GA tracking when clicking edit link button

This will fire an `editButtonClick` action whenever a person visiting an index page decides to edit a link.

Commit:
Adds GA tracking on Edit Link page

This will fire the following events:

- Clicking the Save button: `track-submit` (action: `save-button-click`, category: `userInteraction:LLM`)
- Clicking the Delete button and accepting the "Are you sure?" confirmation message: `track-click` (action: `delete-button-click`, category: 'userInteraction:LLM')
- Getting an error on the form: `auto-track-event` (action: 'alert-error', category: `userAlert:LLM`)

fofr commented:
This advanced analytics tracking is no longer required.
